### ISSUE TITLE
New screenshot web API

### ISSUE THREAD

**Issue by dcmc87** [Mokhet/MFF-roster-tracker#43]
### OCR Web API
I decided to port my OCR app to a web version so it's more useful. Taken a while because I had to learn python, basic html/js, and hosting on an unmanaged VPS from scratch haha, but was a very educational project. There's a lot more room for improvement with the OCR pre-processing, but image processing theory is pretty complicated.

I've written a readme on the project page @ https://github.com/dcmc87/MFFRosterParserWeb, and the POST url is [www.dancmc.io/mff/ocr](www.dancmc.io/mff/ocr). That address is also hosting a functional demo website where you can upload screenshots to test.

I'm thinking you could ideally build a front end UI on your website and just POST requests to my link, it'd probably look much nicer than anything I could make and allow direct integration with the results. 

### Issues
1. One issue with copy and pasting JSON generated by me into your site currently is that it overwrites all character info. For example, if I post character detail data in, it sets all gears to 0, and vice versa. One way around this is to allow importing of JSON fragments, or to have an option to not replace values that are 0 in the JSON.  

2. I realised there are some duplicate gear names between characters, what I've done on my demo page is to display the thumbnail sent back from my server, and ask the user to choose between the possible characters.  

3. If you build your own frontend, you could aggregate all the results after the user has resolved any duplicate issues, and apply it directly. Or even highlight missing values for users to fill in (JSON contains 0.0 or "" if OCR cannot recognise some words/numbers)

I'd be happy to make any changes to my result data structure, just let me know =)

**Comment by Mokhet** [comment]
Wonderfull. I have no time to check out today, but tomorrow week-end I will for sure. Sounds very interesting.

That is an awesome news.

**Comment by Mokhet** [comment]
I lied :) I took a little time to check out and so far it is great. And I **will definitely** use it on the site. 

However I have a few "enhancement requests" to be able to use the API. Do you want us to discuss about enhancements and sync our data here or on your repository ? I think your repository is more dedicated for it, but I will follow your recommendations.

Can't wait to sync my tool and your API, it will be very great.

**Comment by dcmc87** [comment]
Thanks, yes we can discuss it on my repo. And definitely just let me know
what enhancements you need and I'll try to add them, it's my first time
building something like this so I'm sure there are a lot of things to
improve on =)

By the way I'm probably going to have to retrain the font and recalculate
all the bounding rectangles when they completely revamp the UI next update
lol

**Comment by Mokhet** [comment]
Great. I have prepared my suggestions to help me call your API from the site. I am creating "issues" on your repo.

**Comment by Mokhet** [comment]
UI so far.
![ocr-ui-1](https://cloud.githubusercontent.com/assets/24301953/25064638/acfab8fa-21fe-11e7-87eb-7eabb1395670.png)
![ocr-ui-2](https://cloud.githubusercontent.com/assets/24301953/25064640/b251e526-21fe-11e7-8f9e-9f967d1f0e9d.png)

**Comment by dcmc87** [comment]
Looks great! I've finished reconfiguring my api response by the way, Have a look at the readme for the single file mode and ignore the multi file which returns the old response.

**Comment by Mokhet** [comment]
I have no idea about server architecture unfortunately.

**Comment by Mokhet** [comment]
Almost completed. I hope I can finish it by tomorrow evening and let you play with the beta version :D Thanks again, that is a very huge thing to have for the tool.

**Comment by dcmc87** [comment]
I've finished migrating to my new server, but the DNS change has only propagated to about half the world at the moment. You can post requests directly to the IP at http://173.212.216.199/mff/ocr in the meantime

**Comment by dcmc87** [comment]
I'm also working on getting SSL set up again, you might run into warnings for now trying to post to http from your site

**Comment by Mokhet** [comment]
Sorry for the delay, it is almost ready for first tests. Details is ready, gear is on the way.

**Comment by dcmc87** [comment]
Take your time, no rush at all =)

**Comment by Mokhet** [comment]
Details : ready
Gear : ready
Remove all errors : ready
Apply all : in progress

Way more complicated to check everything than I thought at first :p

**Comment by dcmc87** [comment]
Haha features often turn out more complicated than we think. I spent ages
trying to make the base OCR software read it accurately and failing, then I
finally sat down and learnt how to do the custom font training and solved
it all in a day

**Comment by Mokhet** [comment]
Beta test is available online https://www.mokhet.com/MFF-beta/

You should export your data before testing, safer to save it :D

So far, only tested on Safari Mac. I am going to do some tests with Chrome Mac and Firefox Mac later tonight, and tomorrow at work I will be able to do tests on windows for IE11, Edge, Chrome and Firefox ESR. I hope it is working good enough for the first tests.

**Comment by Mokhet** [comment]
Feel free to tell me if anything can be improved/updated. I hope it is simple enough, but you might have a fresher eye on the UI.

**Comment by Mokhet** [comment]
Oops, fixed an issue with gear position 1 (which in fact is position 0 in array and did not fit correctly my checks).

Beside this bug, all my tests are ok for Firefox, chrome and safari Mac. I hope I will have same successes tomorrow on windows 7 and windows 10

**Comment by Mokhet** [comment]
Arghhh and now I find a chrome. grrr I definitely hate this browser.

**Comment by Mokhet** [comment]
Fixed :D Yoohoo, it was not a bug related to OCR, sadly it was already there before :(

**Comment by dcmc87** [comment]
Whoo looks good. I also fixed a bug on my end.

OK some thoughts from my limited testing :
1. Could you put a regular file input multiple button element there as well
in addition to the drag and drop? I often have my browser on Mac running in
full screen and it's quite difficult to do drag and drop

2. It doesn't seem to be doing any file compression client side and just
uploading the full size pics, which is slowing uploads and taking up disk
space =p I wrote a some JS to do compression of image file arrays, let me
know if you want the code

3. In the gear upload result, maybe should let people change the attribute
type instead of just the value? Also for blank results, maybe put a couple
of blank boxes for input or remove the word invalid? Because say my gear
level is +12, it's correct that the last few attributes would be blank, but
putting invalid property may make people think there was an error processing

Other than that looks fantastic for now =)

**Comment by dcmc87** [comment]
Oh yeah the other thing with point 1 is that you can't drag and drop on
iPad/phone, so a button is needed

**Comment by Mokhet** [comment]
1) Of course, what was I thinking about ? Easy fix, will add the multiple browse button under the previews.

2) If you already have a js code for this, I will happily use it instead of digging the internet to find something usable.

3) Once again all valid points, I did not think about characters with gear below 20. I'm so use to track nothing on those characters until they reach 60/20/20/20/20 :)

All are valid points, I will address them tonight when I'll be back from work, but for now I am so very late, time to leave (run) house :D

**Comment by dcmc87** [comment]
Thanks! I'll have the js code up in a separate GitHub project and link you
in, currently working fine but I'm trying to figure out how to use web
workers so the browser can compress multiple images concurrently

**Comment by Mokhet** [comment]
1. issue #46 
2. issue #47
3. issues #48 and #49

**Comment by dcmc87** [comment]
Sorry about the delay, finally got stuff configured the way I wanted. Compression code is at https://github.com/dcmc87/JSImageCompress. Demo is at https://dancmc.io/projects/JSImageCompress

**Comment by Mokhet** [comment]
np :)

Do you mind if I create my own reworked branch ? My linter is not pleased with the code style and global methods :)

**Comment by dcmc87** [comment]
Haha sure go ahead and do whatever you want with it. As long as you send a
blob or base64string my server is happy haha

**Comment by Mokhet** [comment]
All done and live on https://www.mokhet.com/MFF-beta/

For the compression options, I have used { "maxSize" : 1, "speed" : 10 } with default units to "mb"

I hope your server will like it better :p

Tell me if you want me to reduce the maxSize or add another specific option.

**Comment by dcmc87** [comment]
Max size should be 0.1 (100kb images have all worked for me so far) and
speed can probably be around 5 (1 is fastest) though honestly 2-3 would be
fine too

**Comment by Mokhet** [comment]
Done, updated to { "maxSize" : 0.1, "speed" : 3 }

If we happens to have too many fails we will see :D

**Comment by Mokhet** [comment]
So far, all my tests are working very great. This is going to be so simple now, your work is really amazing.

**Comment by dcmc87** [comment]
It's working really well =) But I'm still receiving the full sized image on
the server side, I'm wondering if something went wrong with the compression
code or I'm not loading the new js file or something?

**Comment by dcmc87** [comment]
Ok it's not a problem with cached code, I'm scanning through the
compression code in inspector to see if I can figure out the issue

**Comment by Mokhet** [comment]
Hummm. this probably a mistake on my side. I'll check in half an hour when I'll be back.

-- 
Vilday Laurent

**Comment by dcmc87** [comment]
I found the bug =)

In compress.js, in resultArray.push, you put "file" : file, but that's
pushing the original uncompressed file. I think you meant to push something
like dataUrlToBlob(dataurl) instead haha

**Comment by Mokhet** [comment]
Yes, I just figured out this 15 seconds ago, working on it. Will be ready in a few :) sorry.

**Comment by Mokhet** [comment]
Fixed, should be better. I hope :D

**Comment by dcmc87** [comment]
Yup everything's working perfectly now =) =)

Btw, is there any purpose to the window closing in 5 secs after pressing
apply?

**Comment by Mokhet** [comment]
Great.

Only purpose for the 5 secs delay is to tell to let the user acknowledge everything is complete. At first I was automatically closing the popup and it felt like it was bugged.

**Comment by Mokhet** [comment]
OMG again, a problem with chrome. I so definitely hate this browser.

The image is correctly parsed from safari and firefox but got errorr with chrome (all on Mac).

Do you think it is related to compression ?

- Original screenshot :
![img_3221](https://cloud.githubusercontent.com/assets/24301953/25307326/75e1d6fe-279f-11e7-9f4d-9900aeed8a21.PNG)

- Safari result : good
<img width="610" alt="safari" src="https://cloud.githubusercontent.com/assets/24301953/25307336/b023d7f4-279f-11e7-9b1a-1a3058b55703.png">

- Firefox result : good
<img width="609" alt="firefox" src="https://cloud.githubusercontent.com/assets/24301953/25307375/4dfc659a-27a0-11e7-849a-ccd2c8513d4c.png">

- Chrome result : some mistakes
<img width="607" alt="chrome" src="https://cloud.githubusercontent.com/assets/24301953/25307378/533b523c-27a0-11e7-8657-e8dfb1a08b73.png">

**Comment by dcmc87** [comment]
That's really strange. I had no issues so far with chrome on windows,
safari on Mac and chrome on android. Will test on chrome Mac when I get
home and see what's going on.

It sounds like I may need to tweak something on the server side. Possible
that chrome Mac is using a less optimised jpeg compression so the image
isn't quite good enough, but will have to check

**Comment by dcmc87** [comment]
Had a peek at the server logs on my phone, looks like the Mac chrome image
really is just noisier and introducing random characters. I'll see if I can
replicate it on my macbook at home and tweak the OCR algorithm.

If I can't then we'll just have to increase the maxsize of the compression
a bit

**Comment by dcmc87** [comment]
Yup just as I thought it was just different browser's jpeg compression
affecting quality. Although strangely enough chrome on my macbook worked
with your original image.

 But I tweaked my OCR threshold setting and the details page should be fine
now. Gear page might need a bit more testing to make sure the thresholds
are correctly tuned, but no point trying too much till 3.0 hits

**Comment by Mokhet** [comment]
Probably a version issue in the chrome jpeg algorithm, my chrome was not on auto update and the version was 57.0.2987.133 (64-bit)

Now I updated to 58.0.3029.81 (64-bit) and everything is fine with the exact same image.

**Comment by dcmc87** [comment]
Oh I already updated the server code so it's probably hard to tell. But
either way it's fixed =)

**Comment by Mokhet** [comment]
Great :)

**Comment by Mokhet** [comment]
So, what is our next move ?

1) public release
2) beta public release
3) wait for MFF 3.0 before option 1 or 2
4) alliance release
5) another option

What do you think ?

**Comment by dcmc87** [comment]
I'm leaning towards option 3, if only because I'm hopeful it will be
released this week and then I can rework the OCR. Otherwise if not we can
just go with option 1

**Comment by Mokhet** [comment]
Good, give me free time for a few code tweaking :)

**Comment by Mokhet** [comment]
Just finished reading the patch notes, I saw nothing about the 2 screens we are using, I hope the "new" UI did not messed up with ´em. Only a few hours to wait, first time I can't wait for the update, not for the game update itself but for our own update :D

**Comment by dcmc87** [comment]
Haha fingers crossed, it looks like the font probably didn't change, just hope the pixel coordinates didn't change either

**Comment by dcmc87** [comment]
https://docs.google.com/spreadsheets/d/1Sl3Wm05-WsLIlPe94aMFpjGAtVjH0lmGNnzTVNwrqyE/edit?usp=sharing

By the way, could you help me fill in the corresponding char_alias/uni_alias for 3.0 chars/unis in this sheet? Those are what they will be named by you on your end. I'll type in the gear names when they come out (hope someone posts rocket's new gears cos I'm not planning to buy that uni haha).

**Comment by Mokhet** [comment]
I'll help. Most probably tomorrow evening, after work.

**Comment by Mokhet** [comment]
Done, except for Whiplash since I dunno yet.

**Comment by dcmc87** [comment]
I've already updated the OCR coordinates and it's live, may need to test a
bit though so far looks good. I'll push the database update soon

**Comment by Mokhet** [comment]
I also have updated the characters data in beta directory.

-- 
Vilday Laurent

**Comment by Mokhet** [comment]
I'm planning to do more tests later today when I'll be back from work. If everything is ok on my side, do you give me a go for public release ?

**Comment by dcmc87** [comment]
Yup definitely go ahead. I need to write some automated web tests for my
api later too, I had to manually change all the bounding boxes for each
value in each aspect ratio so pretty sure there's a mistake or two somewhere

**Comment by dcmc87** [comment]
I just discovered a sort of bug. With character gears that are ambiguous
like sharonrogers:4, winter soldier:1, there should be a filter based on
gear attributes - eg if there is def pen, it can't possibly be winter
soldier:1.

It should be easy enough to implement a filter server-side, will work on it

**Comment by Mokhet** [comment]
I have created issue #52 on this matter. I think I can also add some kind of check to prevent such "bug".

Do you have created a related issue on your repository ?

Since I am about to merge the branch 2.4 to master branch, this thread will probably be closed. But we can still communicate on this closed thread I guess :)

**Comment by Mokhet** [comment]
Do you log internal server errors ? I got a them with "Agent venom" gears level 5 gear (absolutely
 not developed :)

How do you want us to work for those kind of errors. I create issue here or on your repository ?

Here are the screens giving internal server errors

![img_3268](https://cloud.githubusercontent.com/assets/24301953/25559648/7bcd7938-2d3f-11e7-9ed5-e169d8aeff4b.PNG)
![img_3269](https://cloud.githubusercontent.com/assets/24301953/25559649/7be91cd8-2d3f-11e7-9ed4-71b757fb89e1.PNG)
![img_3270](https://cloud.githubusercontent.com/assets/24301953/25559650/7be9f310-2d3f-11e7-862e-88c51d047ed0.PNG)

**Comment by Mokhet** [comment]
When gear 1 is not giving error.

![img_3267](https://cloud.githubusercontent.com/assets/24301953/25559659/a5165e40-2d3f-11e7-806b-58cb35c69a53.PNG)

**Comment by Mokhet** [comment]
The mobirum link is http://www.mobirum.com/article/detail?cafeId=futurefight_en&bbsId=84&id=798117

**Comment by dcmc87** [comment]
I think it's fine to create an issue either side, but for this one I'll
create one my side as well. I've isolated the bug, easy enough fix. But in
the process I found a bigger bug, where I forgot that at gear +5 there is
only one fixed orange stat =( Working on that now
